<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#投票法" data-toc-modified-id="投票法-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>投票法</a></span><ul class="toc-item"><li><span><a href="#融合模型" data-toc-modified-id="融合模型-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>融合模型</a></span></li><li><span><a href="#构建多样性" data-toc-modified-id="构建多样性-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>构建多样性</a></span></li><li><span><a href="#分类器加权" data-toc-modified-id="分类器加权-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>分类器加权</a></span></li></ul></li><li><span><a href="#堆叠法Stacking" data-toc-modified-id="堆叠法Stacking-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>堆叠法Stacking</a></span><ul class="toc-item"><li><span><a href="#细节" data-toc-modified-id="细节-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>细节</a></span></li><li><span><a href="#元学习器的特征矩阵" data-toc-modified-id="元学习器的特征矩阵-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>元学习器的特征矩阵</a></span></li><li><span><a href="#样本量太少的解决方案：交叉验证" data-toc-modified-id="样本量太少的解决方案：交叉验证-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>样本量太少的解决方案：交叉验证</a></span></li><li><span><a href="#特征太少的解决方案" data-toc-modified-id="特征太少的解决方案-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>特征太少的解决方案</a></span></li><li><span><a href="#Stacking融合的训练和测试流程" data-toc-modified-id="Stacking融合的训练和测试流程-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Stacking融合的训练和测试流程</a></span><ul class="toc-item"><li><span><a href="#stacking的训练:" data-toc-modified-id="stacking的训练:-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span><strong>stacking的训练</strong>:</a></span></li><li><span><a href="#stacking的测试" data-toc-modified-id="stacking的测试-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>stacking的测试</a></span></li></ul></li></ul></li><li><span><a href="#Blending混合法" data-toc-modified-id="Blending混合法-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Blending混合法</a></span></li><li><span><a href="#装袋算法Bagging" data-toc-modified-id="装袋算法Bagging-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>装袋算法Bagging</a></span><ul class="toc-item"><li><span><a href="#优缺点" data-toc-modified-id="优缺点-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>优缺点</a></span></li><li><span><a href="#OOB策略" data-toc-modified-id="OOB策略-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>OOB策略</a></span></li></ul></li></ul></div>

**集成学习 (Ensemble Learning)** 通过构建并结合多个学习器来完成学习任务，有时也被称作多分类器系统 (Multi-classifier System)、基于委员会的学习 (Committee-based Learning)等。

《机器学习》- 周志华

**统计方面**：假设空间不同，泛化能力不同
- 集成算法(狭义)：一维假设空间

>每个算法都是一套独一无二的、从数据到结论的计算流程，这一流程在统计学上被称为是“对样本与标签关系的假设”。例如，KNN假设，相同类别的样本在特征上应该相似。Boosting算法假设令单一损失函数最小的树一定可以组成令整体损失最小的树林。
 
>我们可以有不同的假设，但在有限的训练数据集上，我们很难看出哪个假设才是对贴近于真实样本标签关系的假设。这就是说，我们无法保证现在表现好的算法在未知数据集上表现也一定会好。因此使用单一假设/单一算法的缺点就很明显：虽然在当前数据集上当前假设表现最好，但如果其他假设才更贴近数据真实的情况，那当前选择出的模型的泛化能力在未来就无法得到保证了。
 
- 模型融合：高维假设空间

>相对的，如果能够将不同的假设结合起来，就可以降低选错假设的风险。最贴近数据真实状况的假设，即便在当前数据集上表现并不是最好的，但在融合之后可以给与融合的结果更强大的泛化能力。
 
>同时，假设真正的数据规律并不能被现在的数据代表，那融合多个假设也能够拓展假设空间，不同假设最终能够更接近真实规律的可能性也越大。
因此，从统计学的角度来说，模型融合能够：(1) 降低选错假设导致的风险、(2)提升捕捉到真正数据规律的可能性、(3)提升具有更好泛化能力的可能性。


**计算方面**：起点数量不同，模型表现不同
- 集成算法(狭义)：只有一组起点，最终只能得到一个局部最小值

>许多算法在搜索时都会陷入局部最优解，即便有足够多的训练数据，我们也无法判断能否找到理论上真实的最小值。一般在算法训练时，我们把搜索出的局部最小值当作是真实最小值的近似值，但我们却无法估计真实最小值与局部最小值之间有大多的差异。但由于我们只有一个局部最小值，真实最小值在哪里完全没有头绪。
 
- 模型融合：多组不同的起点，最终得到一个局部最小值的范围

>模型融合过程中，每个算法都会有自己的起点、并最终获得多个不同的局部最小值。虽然这些局部最小值当中，没有任意一个等于真正的最小值，但真正的最小值很大可能就在这些局部最小值构成的范围当中。通过结合局部最小值，可以获得更接近真实最小值的近似值，从而提升模型的精度。

因此，从计算方面来看，模型融合能够：(1)降低选错起点假设的风险、(2)降低局部最小值远远偏离真实值的风险、(3)提升算法结果更接近真实最小值的可能性。


<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20221101183239483.png" alt="image-20221101183239483" style="zoom:50%;" />

In [7]:
# 基础数据科学运算库
import numpy as np
import pandas as pd

# 可视化库
import seaborn as sns
import matplotlib.pyplot as plt

# 时间模块
import time

import warnings
warnings.filterwarnings('ignore')

# sklearn库
# 数据预处理
from sklearn import preprocessing
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder

# 实用函数
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, roc_auc_score
from sklearn.model_selection import train_test_split

# 常用评估器
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

# 网格搜索
from sklearn.model_selection import GridSearchCV

# 自定义评估器支持模块
from sklearn.base import BaseEstimator, TransformerMixin, ClassifierMixin

# 投票法

适用于分类任务，对多个学习器的预测结果进行投票，即少数服从多数。

投票法有两种：**普通投票法和加权投票法**。

加权的权重可以人工主观设置或者根据模型评估分数来设置权重。**投票需要3个及3个以上的模型**，同时建议要保证模型的多样性，有时候对同质模型们使用投票法并不能取得较好的表现，这是因为同质模型得到的结果之间可能具有较强的相关性，从而会导致多数人把少数人的好想法给压下去了。

为了避免这个问题，可以参考在2014年KDD Cup上Marios Michailid的做法，他对所有结果文件计算Pearson系数，最后选取其中相关性小的模型结果进行投票，分数获得了提升。

**投票法对所有评估器输出的结果按类别进行计数，出现次数最多的类别就是融合模型输出的类别**

* 绝对多数投票法：至少有50%或以上的分类器都输出了同一类别，否则就拒绝预测, (否则则输出“不确定”，相当于增加一类输出类别)。

>绝对多数投票可以帮助我们衡量当前投票的置信程度。票数最多的类别占比越高，说明融合模型对当前样本的预测越有信心。当票数最多的类别占比不足50%时，重新投票对于机器学习来说效率太低，因此我们往往会选择另一种方案：我们可以使用相对较弱的学习器进行绝对多数投票融合、并将绝对多数投票中被输出为“不确定”的样本交由学习能力更强的模型进行预测，相当于将“困难样本”交由复杂度更高的算法进行预测，这也是一种融合方式。

* 相对多数投票法：最终结果在投票中票数最多。(**少数服从多数**)

* 加权投票法：加权的权重可以人工主观设置或者根据模型评估分数来设置权重。

* 硬投票：对多个模型直接进行投票，不区分模型结果的相对重要度，最终投票数最多的类为最终被预测的类。

* 软投票：增加了设置权重的功能，可以为不同模型设置不同权重，进而区别模型不同的重要度。

>软投票和硬投票得出的结果可能不同，这是因为软投票有一种“窥探心灵”的能力（也就是衡量置信度的能力）。

>在实际进行投票法融合时，我们往往优先考虑软投票 + 相对多数投票方案，因为软投票方案更容易在当前数据集上获得好的分数。然而，如果我们融合的算法都是精心调参过的算法，那软投票方案可能导致过拟合。因此具体在使用时，需要依情况而定。

## 融合模型

```Voting```可通过```VotingClassifier```和```VotingRegressor```使用。可以将基本模型列表作为参数，列表中的每个模型都必须是具有名称和模型的元组

<center><img src="https://s2.loli.net/2022/05/20/qY5UI3gDzSc8p2H.png" alt="image-20220520153651685" style="zoom:33%;" />

根据少数如从多数的原则，投票得出三个样本的预测结果分别为1、0、1、0。

把这个**少数服从多数的过程等价转化是否超过半数评估器认为该样本应该属于1类**，如果是，则输出结果为1，反之则输出预测结果为0。需要注意的是，该做法会更加方便代码层面的实现，也是后续我们主要采用的计算流程。

<center><img src="https://s2.loli.net/2022/05/20/uVrzS79LaBsJgKp.png" alt="image-20220520122231347" style="zoom:33%;" />

在单体分类器准确率为80%左右（较为普遍的情况）时，模型投票融合能有平均约15%的准确率提升。当然，该理论实际上是基于分类器相互独立的假设推导而来，而在大多数真实场景下，该假设并不成立，因此该理论的结论可以视作一个理论上限，并不能代表一般情况。

In [9]:
from sklearn.ensemble import VotingClassifier

评估器核心参数解释如下：
|Name|Description|   
|:--:|:--:| 
|estimators|由（评估器名称，评估器）所组成的列表|
|voting|融合的方式，包括此前介绍的硬投票和软投票两种{‘hard’,‘soft’}, default=’hard’|
|weights|融合过程中各评估器的权重|
|flatten_transform|在软投票时打印结果方式|

In [2]:
#常用工具库
import re
import numpy as np
import pandas as pd
import matplotlib as mlp
import matplotlib.pyplot as plt
import time
 
#算法辅助 & 数据
import sklearn
from sklearn.model_selection import KFold, cross_validate
from sklearn.datasets import load_digits #分类 - 手写数字数据集
from sklearn.model_selection import train_test_split
 
#算法（单一学习器）
from sklearn.neighbors import KNeighborsClassifier as KNNC
from sklearn.neighbors import KNeighborsRegressor as KNNR
from sklearn.tree import DecisionTreeRegressor as DTR
from sklearn.tree import DecisionTreeClassifier as DTC
from sklearn.linear_model import LinearRegression as LR
from sklearn.linear_model import LogisticRegression as LogiR
from sklearn.ensemble import RandomForestRegressor as RFR
from sklearn.ensemble import RandomForestClassifier as RFC
from sklearn.ensemble import GradientBoostingRegressor as GBR
from sklearn.ensemble import GradientBoostingClassifier as GBC
from sklearn.naive_bayes import GaussianNB
import xgboost as xgb
 
#融合模型
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import VotingRegressor

In [6]:
def individual_estimators(estimators):
    """
    对模型融合中每个评估器做交叉验证，对单一评估器的表现进行评估
    """
    for estimator in estimators:
        cv = KFold(n_splits=5,shuffle=True,random_state=1412)
        results = cross_validate(estimator[1],Xtrain,Ytrain
                             ,cv = cv
                             ,scoring = "accuracy"
                             ,n_jobs = -1
                             ,return_train_score = True
                             ,verbose=False)
        test = estimator[1].fit(Xtrain,Ytrain).score(Xtest,Ytest)
        print(estimator[0]
          ,"\n train_score:{}".format(results["train_score"].mean())
          ,"\n cv_mean:{}".format(results["test_score"].mean())
          ,"\n test_score:{}".format(test)
          ,"\n")
        
def fusion_estimators(clf):
    """
    对融合模型做交叉验证，对融合模型的表现进行评估
    """
    cv = KFold(n_splits=5,shuffle=True,random_state=1412)
    results = cross_validate(clf,Xtrain,Ytrain
                             ,cv = cv
                             ,scoring = "accuracy"
                             ,n_jobs = -1
                             ,return_train_score = True
                             ,verbose=False)
    test = clf.fit(Xtrain,Ytrain).score(Xtest,Ytest)
    print("train_score:{}".format(results["train_score"].mean())
          ,"\n cv_mean:{}".format(results["test_score"].mean())
          ,"\n test_score:{}".format(test)
         )

* train_score：模型在训练集上进行训练的得分表现的均值
* cv_mean：模型通过交叉验证在验证集上的得分表现的均值
* test_score：训练好的模型在测试集上的得分表现

In [4]:
data = load_digits()     # 使用sklearn自带的手写数字数据集，是一个10分类数据集。

X = data.data
y = data.target

Xtrain,Xtest,Ytrain,Ytest = train_test_split(X,y,test_size=0.2,random_state=1412)

**第一组分类器**：放任自由，收敛为主，有较高过拟合风险
>逻辑回归+随机森林+梯度提升树

In [7]:
# 构建多组分类器、进行融合
clf1 = LogiR(max_iter = 3000,random_state=1412,n_jobs=8)
clf2 = RFC(n_estimators= 100,random_state=1412,n_jobs=8)
clf3 = GBC(n_estimators= 100,random_state=1412)
 
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2), ("GBDT",clf3)]
clf = VotingClassifier(estimators,voting="soft")

# 模型评估
individual_estimators(estimators)
fusion_estimators(clf)

Logistic Regression 
 train_score:1.0 
 cv_mean:0.9666061749903212 
 test_score:0.9527777777777777 

RandomForest 
 train_score:1.0 
 cv_mean:0.9763574332171892 
 test_score:0.9833333333333333 

GBDT 
 train_score:1.0 
 cv_mean:0.9686943476577623 
 test_score:0.9388888888888889 

train_score:1.0 
 cv_mean:0.9777390631049168 
 test_score:0.9805555555555555


**通过集成模型的表现有了显著提升，交叉验证分数（97.77%）比任何单一学习器都高，且在测试集上的表现提升到了98.05%，可见当前数据下模型融合的效果十分显著**

并不是任意评估器、任意数据上都能够看到如此一目了然的结果。

如果你进行投票或平均融合之后，融合的结果反而没有单个算法好，那你可能是落入了投票法与平均法会失效的几大陷阱：

---
* **评估器之间的学习能力/模型表现差异太大**。在融合中，一个表现很差的模型会拉低整个融合模型的上限，尤其是回归类算法，当一个模型的表现很差时，平均法得出的结果很难比最好的单一算法还好。因此我们必须要使用表现相似的模型进行融合。如果你的评估器中有拖后腿的模型，无论这个模型有多么先进，都应该立刻把它剔除融合模型。
 ---
* **评估器在类型上太相似**，比如、全是树模型、都是Boosting算法，或都是线性评估器等。如果评估器类别太相似，模型融合会发挥不出作用，这在直觉上其实很好理解：如果平均/投票的评估器都一致，那融合模型最终得出的结果也会与单个评估器一致。
 ---
* **对评估器进行了过于精密的调优**。一般来说，我们可能会认为，先对模型进行调优后再融合，能够进一步提升模型的表现。经过粗略调优的评估器融合确实能提升模型表现，但如果对评估器进行过于精密的调优，可能会让融合后的算法处于**严重过拟合**的状态。因此，一般我们不会在评估器上进行太精准的调优。


---

**第二组分类器**：略微调参（非精细化调参），限制过拟合

> 模型融合是一个可能加剧过拟合的手段，因此我们必须保证每一个学习器本身的过拟合不严重，为此我们需要对模型进行抗过拟合的处理。需要注意的是，对抗过拟合可能会削弱模型的预测效果，因此我们必须根据过拟合的情况、泛化能力的展现来进行选择。

> 对于逻辑回归，我们需要缩小参数C，对随机森林我们选择max_depth，对GBDT我们则选择max_features。


In [14]:
clf1 = LogiR(max_iter = 3000, C=0.1, random_state=1412,n_jobs=8) #这一组合可能说明我们的max_iter设置得太大了
clf2 = RFC(n_estimators= 100,max_depth=12,random_state=1412,n_jobs=8)
clf3 = GBC(n_estimators= 100,max_features="sqrt",random_state=1412)
 
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2), ("GBDT",clf3)]
clf = VotingClassifier(estimators,voting="soft")

# 模型评估
individual_estimators(estimators)
fusion_estimators(clf)

Logistic Regression 
 train_score:1.0 
 cv_mean:0.9666085946573751 
 test_score:0.9638888888888889 

RandomForest 
 train_score:1.0 
 cv_mean:0.9763574332171892 
 test_score:0.9833333333333333 

GBDT 
 train_score:1.0 
 cv_mean:0.9777342237708091 
 test_score:0.9722222222222222 

train_score:1.0 
 cv_mean:0.9777463221060783 
 test_score:0.9861111111111112


根据评估结果:
- **GBDT的过拟合调整是非常成功的，交叉验证分数上升的同时，测试集上的分数也得到了大幅提升。**
- **Logistic回归**的过拟合调整不明显,交叉验证的分数并未得到明显提升，但测试集上的结果上升，**说明模型的泛化能力变得更加稳定**
- **随机森林**的过拟合更加严重，应该取消这种过拟合限制

## 构建多样性

在模型融合当中，独立性被称为“多样性”（diversity），评估器之间的差别越大、彼此之间就越独立，因此评估器越多样，独立性就越强。完全独立的评估器在现实中几乎不可能实现，因为不同的算法执行的是相同的预测任务，更何况大多数时候算法们都在相同的数据上训练，因此评估器不可能完全独立。但我们有以下关键的手段，用来让评估器变得更多样、让评估器之间相对独立：

- 训练数据多样性：完成多组有效的特征工程，使用不同的特征矩阵训练不同的模型。该方法一般能够得到很好的效果，但如何找出多组有效的特征工程是难题。
 ---
- 样本多样性：使用相同特征矩阵，但每次训练时抽样出不同的样本子集进行训练。当数据量较小时，抽样样本可能导致模型效果急剧下降。
 ---
- 特征样性：使用相同特征矩阵，但每次训练时抽样出不同的特征子集进行训练。当特征量较小时，抽样特征可能导致模型效果急剧下降。
 ---
- 随机多样性/训练多样性：使用相同的算法，但使用不同的随机数种子（会导致使用不同的特征、样本、起点）、或使用不同的损失函数、使用不同的不纯度下降量等。这一方法相当于是在使用Bagging集成。
 ---
- 算法多样性：增加类型不同的算法，如集成、树、概率、线性模型相混合。但需要注意的是，模型的效果不能太糟糕，无论是投票还是平均法，如果模型效果太差，可能大幅度降低融合的结果。
---
>增加多样性的操作或多或少都阻止了模型学习完整的数据，因此会削弱模型对数据的学习，可能降低模型的效果。因此我们使用多样性时，需要时刻关注着模型的结果。


In [15]:
#逻辑回归没有增加多样性的选项
clf1 = LogiR(max_iter = 3000, C=0.1, random_state=1412,n_jobs=8)
#增加特征多样性与样本多样性
clf2 = RFC(n_estimators= 100,max_features="sqrt",max_samples=0.9, random_state=1412,n_jobs=8)
#特征多样性，稍微上调特征数量
clf3 = GBC(n_estimators= 100,max_features=16,random_state=1412) 
 
#增加算法多样性，新增决策树、KNN、贝叶斯
clf4 = DTC(max_depth=8,random_state=1412)
clf5 = KNNC(n_neighbors=10,n_jobs=8)
clf6 = GaussianNB()
 
#新增随机多样性，相同的算法更换随机数种子
clf7 = RFC(n_estimators= 100,max_features="sqrt",max_samples=0.9, random_state=4869,n_jobs=8)
clf8 = GBC(n_estimators= 100,max_features=16,random_state=4869)
 
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2)
              , ("GBDT",clf3), ("Decision Tree", clf4), ("KNN",clf5) 
              , ("Bayes",clf6), ("RandomForest2", clf7), ("GBDT2", clf8)
             ]
clf = VotingClassifier(estimators,voting="soft")

# 模型评估
individual_estimators(estimators)
fusion_estimators(clf)

Logistic Regression 
 train_score:1.0 
 cv_mean:0.9666085946573751 
 test_score:0.9638888888888889 

RandomForest 
 train_score:1.0 
 cv_mean:0.9735699767711964 
 test_score:0.9777777777777777 

GBDT 
 train_score:1.0 
 cv_mean:0.9777414827719705 
 test_score:0.9722222222222222 

Decision Tree 
 train_score:0.9509369962538313 
 cv_mean:0.8552821331784747 
 test_score:0.8527777777777777 

KNN 
 train_score:0.9832990502137966 
 cv_mean:0.9763429152148664 
 test_score:0.9833333333333333 

Bayes 
 train_score:0.8295060354940024 
 cv_mean:0.8086285327138987 
 test_score:0.7888888888888889 

RandomForest2 
 train_score:1.0 
 cv_mean:0.9749491869918699 
 test_score:0.9833333333333333 

GBDT2 
 train_score:1.0 
 cv_mean:0.9770446186604724 
 test_score:0.975 

train_score:1.0 
 cv_mean:0.9826098528842431 
 test_score:0.9805555555555555


贝叶斯在训练集上的分数很低，这说明模型的学习能力不足

因此**剔除表现不良的算法**,并且精简算法多样性

In [16]:
estimators = [("Logistic Regression",clf1)
              , ("RandomForest", clf2)
              , ("GBDT",clf3)
              #, ("Decision Tree", clf4)
              , ("KNN",clf5) 
              #, ("Bayes",clf6)
              , ("RandomForest2", clf7)
              #, ("GBDT2", clf8)
             ]
clf = VotingClassifier(estimators,voting="soft")

fusion_estimators(clf)

train_score:1.0 
 cv_mean:0.9833067169957413 
 test_score:0.9888888888888889


从模型的稳定性来考虑，还是包含随机多样性的组合更好，但为了更快的运算速率，我们可以使用精简多样性继续往下计算。

## 分类器加权

**第一种选项：使用各个模型交叉验证结果本身作为权重，有过拟合风险**

In [20]:
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2)
              , ("GBDT",clf3), ("Decision Tree", clf4), ("KNN",clf5)]
 
clf_weighted = VotingClassifier(estimators,voting="soft",weights=[0.96660,0.97357,0.97774,0.85528,0.97634])

fusion_estimators(clf_weighted)

train_score:1.0 
 cv_mean:0.9826098528842431 
 test_score:0.9888888888888889


模型在交叉验证上的效果提升了,并且模型没有陷入过拟合

**第二种选项：加大效果好的算法的权重，减小效果差的算法的权重。**

In [19]:
clf_weighted = VotingClassifier(estimators,voting="soft",weights=[0.95,0.95,0.95,0.85,1.2]) #增大
fusion_estimators(clf_weighted)

train_score:1.0 
 cv_mean:0.9833067169957413 
 test_score:0.9888888888888889


过拟合开始发生了，测试集上的结果开始降低。因此我们选择不加大效果好的算法的权重

模型达到了目前为止，在5折交叉验证及测试分数上的最高值。

现在我们给与决策树的权重非常小，说明现在决策树只是在提供多样性，在实际预测方面做出的贡献较少。提供多样性可以让模型的泛化能力增强，让实际预测方面的贡献变小又可以降低决策树本身较低的预测分数带来的影响。至此，我们就得到了一个很好的融合结果。

---

# 堆叠法Stacking

作为强学习器的融合方法，Stacking集模型**效果好、可解释性强、适用复杂数据**三大优点于一身，属于融合领域最为实用的先驱方法

Stacking结构中有两层算法串联，**第一层叫做level 0，第二层叫做level 1**

level 0里面可能包含1个或多个强学习器，而level 1只能包含一个学习器。

在训练中，数据会先被输入level 0进行训练，训练完毕后，level 0中的每个算法会输出相应的预测结果。我们将这些预测结果拼凑成新特征矩阵，再输入level 1的算法进行训练。融合模型最终输出的预测结果就是level 1的学习器输出的结果


<center><img src="photos\192b3e5666f04c5d8b563b8484ccdd95.png" alt="image-20220520122231347" style="zoom:83%;" />

- **level 0**上训练的多个强学习器被称为**基学习器(base-model)**，也叫做个体学习器。

- 在**level 1**上训练的学习器叫**元学习器（meta-model）**。

>根据行业惯例，level 0上的学习器是复杂度高、学习能力强的学习器，例如集成算法、支持向量机 (level 0上的算法们的职责是**找出原始数据与标签的关系、即建立原始数据与标签之间的假设，因此需要强大的学习能力**)

>而level 1上的学习器是可解释性强、较为简单的学习器，如决策树、线性回归、逻辑回归等 (level 1上的算法的职责是**融合个体学习器做出的假设、并最终输出融合模型的结果**，相当于在寻找“最佳融合规则”，而非直接建立原始数据与标签之间的假设)


**本质上，level 1的算法只是在学习如何将level 0上输出的结果更好地结合起来，所以Stacking是通过训练学习器来融合学习器结果的方法**

## 细节
---
- 要不要对融合的算法进行精密的调参？

>**个体学习器粗调，元学习器精调**，如果不过拟合的话，可以两类学习器都精调。理论上来说，算法输出结果越接近真实标签越好，但个体学习器精调后再融合，很容易过拟合。

- 个体学习器算法要怎样选择才能最大化stacking的效果？

>与投票、平均的状况一致，**控制过拟合、增加多样性、注意算法整体的运算时间**。

- 个体学习器可以是逻辑回归、决策树这种复杂度较低的算法吗？元学习器可以是xgboost这种复杂度很高的算法吗？

>都可以，一切以模型效果为准。对level 0而言，当增加弱学习器来增加模型多样性、且弱学习器的效果比较好时，可以保留这些算法。对level 1而言，只要不过拟合，可以使用任何算法。个人推荐，**在分类的时候可以使用复杂度较高的算法，对回归最好还是使用简单的算法**。

- level 0和level 1的算法可不可以使用不同的损失函数？

>可以，因为不同的损失函数衡量的其实是类似的差异：即真实值与预测值之间的差异。不过不同的损失对于差异的敏感性不同，如果可能的话建议使用相似的损失函数。

- level 0和level 1的算法可不可以使用不同的评估指标？

>建议**level 0与level 1上的算法必须使用相同的模型评估指标**。

>虽然Stacking中串联了两组算法，但这两组算法的训练却是完全分离的。在深度学习当中，我们也有类似的强大算法串联弱小算法的结构，例如，卷积神经网络就是由强大的卷积层与弱小的线性层串联，卷积层的主要职责是找出特征与标签之间的假设，而线性层的主要职责是整合假设、进行输出。但在深度学习中，一个网络上所有层的训练是同时进行的，每次降低损失函数时需要更新整个网络上的权重。但**在Stacking当中，level 1上的算法在调整权重时，完全不影响level 0的结果**，因此为了保证两组算法最终融合后能够得到我们想要的结果，在训练时一定要以唯一评估指标为基准进行训练。

In [ ]:
class sklearn.ensemble.StackingClassifier(estimators, final_estimator=None, *, cv=None, stack_method='auto', n_jobs=None, passthrough=False, verbose=0)

<center><img src="photos\6ce51ea0fcdc471792e90e7c3ef0525a.png" alt="image-20220520122231347" style="zoom:83%;" />

In [23]:
#常用工具库
import re
import numpy as np
import pandas as pd
import matplotlib as mlp
import matplotlib.pyplot as plt
import time
 
#算法辅助 & 数据
import sklearn
from sklearn.model_selection import KFold, cross_validate
from sklearn.datasets import load_digits #分类 - 手写数字数据集
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
 
#算法（单一学习器）
from sklearn.neighbors import KNeighborsClassifier as KNNC
from sklearn.neighbors import KNeighborsRegressor as KNNR
from sklearn.tree import DecisionTreeRegressor as DTR
from sklearn.tree import DecisionTreeClassifier as DTC
from sklearn.linear_model import LinearRegression as LR
from sklearn.linear_model import LogisticRegression as LogiR
from sklearn.ensemble import RandomForestRegressor as RFR
from sklearn.ensemble import RandomForestClassifier as RFC
from sklearn.ensemble import GradientBoostingRegressor as GBR
from sklearn.ensemble import GradientBoostingClassifier as GBC
from sklearn.naive_bayes import GaussianNB
import xgboost as xgb
 
#融合模型
from sklearn.ensemble import StackingClassifier

In [24]:
#逻辑回归没有增加多样性的选项
clf1 = LogiR(max_iter = 3000, C=0.1, random_state=1412,n_jobs=8)
#增加特征多样性与样本多样性
clf2 = RFC(n_estimators= 100,max_features="sqrt",max_samples=0.9, random_state=1412,n_jobs=8)
#特征多样性，稍微上调特征数量
clf3 = GBC(n_estimators= 100,max_features=16,random_state=1412) 
 
#增加算法多样性，新增决策树与KNN
clf4 = DTC(max_depth=8,random_state=1412)
clf5 = KNNC(n_neighbors=10,n_jobs=8)
clf6 = GaussianNB()
 
#新增随机多样性，相同的算法更换随机数种子
clf7 = RFC(n_estimators= 100,max_features="sqrt",max_samples=0.9, random_state=4869,n_jobs=8)
clf8 = GBC(n_estimators= 100,max_features=16,random_state=4869)
 
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2)
              , ("GBDT",clf3), ("Decision Tree", clf4), ("KNN",clf5) 
              #, ("Bayes",clf6)
              , ("RandomForest2", clf7), ("GBDT2", clf8)
             ]

#选择单个评估器中分数最高的随机森林作为元学习器
#也可以尝试其他更简单的学习器
final_estimator = RFC(n_estimators=100
                      , min_impurity_decrease=0.0025
                      , random_state= 420, n_jobs=8)

clf = StackingClassifier(estimators=estimators #level0的7个体学习器
                         ,final_estimator=final_estimator #level 1的元学习器
                         ,n_jobs=8)

# 模型评估
individual_estimators(estimators)
fusion_estimators(clf)

Logistic Regression 
 train_score:1.0 
 cv_mean:0.9666085946573751 
 test_score:0.9638888888888889 

RandomForest 
 train_score:1.0 
 cv_mean:0.9735699767711964 
 test_score:0.9777777777777777 

GBDT 
 train_score:1.0 
 cv_mean:0.9777414827719705 
 test_score:0.9722222222222222 

Decision Tree 
 train_score:0.9509369962538313 
 cv_mean:0.8552821331784747 
 test_score:0.8527777777777777 

KNN 
 train_score:0.9832990502137966 
 cv_mean:0.9763429152148664 
 test_score:0.9833333333333333 

RandomForest2 
 train_score:1.0 
 cv_mean:0.9749491869918699 
 test_score:0.9833333333333333 

GBDT2 
 train_score:1.0 
 cv_mean:0.9770446186604724 
 test_score:0.975 

train_score:1.0 
 cv_mean:0.9805216802168022 
 test_score:0.9888888888888889


**数据学习难度较大时**，stacking的优势就会慢慢显现出来。

我们现在使用的元学习器几乎是默认参数，我们可以针对元学习器使用贝叶斯优化进行精妙的调参，然后再进行对比，堆叠法的效果可能超越投票法

## 元学习器的特征矩阵

在Stacking过程中，个体学习器会原始数据上训练、预测，再把预测结果排布成新特征矩阵，放入元学习器进行学习。其中，个体学习器的预测结果、即元学习器需要训练的矩阵一般如下排布：

<center><img src="photos\93a360e35bd94f8da142b221d3f62695.png" alt="image-20220520122231347" style="zoom:83%;" />

**新特征矩阵中的特征数就等于个体学习器的个数**。

>一般融合模型中个体学习器最多有20-30个，也就是说元学习器的特征矩阵中最多也就20-30个特征。这个特征量对于工业、竞赛中的机器学习算法来说是远远不够的

在我们训练stacking模型时，我们一定是将原始数据集分为训练集、验证集和测试集三部分：

其中测试集是用于检测整个融合模型的效果的，因此在训练过程中不能使用。

而训练集用于训练个体学习器，属于已经完全透露给个体学习器的内容，**如果在训练集上进行预测，那预测结果是“偏高”的、无法代表个体学习器的泛化能力**。

因此最后剩下能够用来预测、还能代表个体学习器真实学习水平的，就**只剩下很小的验证集了**。
>一般验证集最多只占整个数据集的30%-40%，这意味着元学习器所使用的特征矩阵里的样本量最多也就是原始数据的40%。


## 样本量太少的解决方案：交叉验证

**参数cv，在stacking中执行交叉验证**

在stacking方法被提出的原始论文当中，原作者自然也意识到了元学习器的特征矩阵样本量太少这个问题，因此提出了**在stacking流程内部使用交叉验证来扩充元学习器特征矩阵的想法**，即在内部对每个个体学习器做交叉验证，但并不用这个交叉验证的结果来验证泛化能力，而是直接把交叉验证当成了生产数据的工具。


具体的来看，在stacking过程中，我们是这样执行交叉验证的，对任意个体学习器来说，假设我们执行5折交叉验证，我们会将训练数据分成5份，并按照4份训练、1份验证的方式总共建立5个模型，训练5次：


<center><img src="photos\997d2ab86f2b44bca2761d5c52a66cf3.png" alt="image-20220520122231347" style="zoom:83%;" />

交叉验证的最终输出是5个验证集上的分数，但计算分数之前我们一定是在5个验证集上分别进行预测，并输出了结果。

所以我们可以在交叉验证中建立5个模型，轮流得到5个模型输出的预测结果，而这5个预测结果刚好对应全数据集中分割的5个子集。这是说，我们完成交叉验证的同时，也对原始数据中全部的数据完成了预测


**将5个子集的预测结果纵向堆叠，就可以得到一个和原始数据中的样本一一对应的预测结果**

In [28]:
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2)
              , ("GBDT",clf3), ("Decision Tree", clf4), ("KNN",clf5) 
              #, ("Bayes",clf6)
              , ("RandomForest2", clf7), ("GBDT2", clf8)
             ]
final_estimator = RFC(n_estimators=100
                      , min_impurity_decrease=0.0025
                      , random_state= 420, n_jobs=8)

In [29]:
def cvtest(cv):
    clf = StackingClassifier(estimators=estimators
                         ,final_estimator=final_estimator
                         , cv = cv
                         , n_jobs=8)
    start = time.time()
    clf.fit(Xtrain,Ytrain)
    print((time.time() - start)) #消耗时间
    print(clf.score(Xtrain,Ytrain)) #训练集上的结果
    print(clf.score(Xtest,Ytest)) #测试集上的结果

In [30]:
cvtest(2)     # 较少的验证次数

6.755136013031006
1.0
0.9861111111111112


In [31]:
cvtest(10)

19.262344121932983
1.0
0.9833333333333333


In [32]:
cvtest(30)   # 很大的验证次数

52.01742649078369
1.0
0.9833333333333333


## 特征太少的解决方案
---
参数```stack_method```，更换个体学习器输出的结果类型

* 对于分类stacking来说，如果特征量太少，我们可以更换个体学习器输出的结果类型。具体来说，如果个体学习器输出的是具体类别（如[0,1,2]），那1个个体学习器的确只能输出一列预测结果。

但如果**把输出的结果类型更换成概率值、置信度等内容**，输出结果的结构一下就可以从一列拓展到多列。

这个行为由参数stack_method控制，这是只有StackingClassifier才拥有的参数，它控制个体分类器具体的输出。

```stack_method```里面可以输入四种字符串：```"auto", "predict_proba", "decision_function", "predict"```，除了"auto"之外其他三个都是sklearn常见的接口。


> predict_proba：输出概率值;

* 对二分类，输出样本的真实标签1的概率，一列
* 对n分类，输出样本的真实标签为[0,1,2,3...n]的概率，一共n列

> decision_function：每个样本点到分类超平面的距离，可以衡量置信度;对于无法输出概率的算法，如SVM，我们通常它来输出置信度

* 对二分类，输出样本的真实标签为1的置信度，一列
* 对n分类，输出样本的真实标签为[0,1,2,3...n]的置信度，一共n列

> predict：输出具体的预测标签

* 对任意分类形式，输出算法在样本上的预测标签，一列

**对参数stack_method有**：

>输入"auto"，sklearn会在每个个体学习器上按照"predict_proba", "decision_function", "predict"的顺序，分别尝试学习器可以使用哪个接口进行输出。即，如果一个算法可以使用predict_proba接口，那就不再尝试后面两个接口，如果无法使用predict_proba，就尝试能否使用decision_function。

输入三大接口中的任意一个接口名，则默认全部个体学习器都按照这一接口进行输出。然而，如果遇见某个算法无法按照选定的接口进行输出，stacking就会报错。


对于回归类算法，我们没有这么多可以选择的接口。回归类算法的输出永远就只有一列连续值

**因而我们可以考虑将原始特征矩阵加入个体学习器的预测值，构成新特征矩阵**。
>(这个操作有较高的过拟合风险，因此当特征过于少、且stacking算法的效果的确不太好的时候，我们才会考虑这个方案。)

控制是否将原始数据加入特征矩阵的参数是```passthrough```。

>设置为True时，则将原始特征矩阵加入个体学习器的预测值、构成大特征矩阵。


In [33]:
estimators = [("Logistic Regression",clf1), ("RandomForest", clf2)
              , ("GBDT",clf3), ("Decision Tree", clf4), ("KNN",clf5) 
              #, ("Bayes",clf6)
              , ("RandomForest2", clf7), ("GBDT2", clf8)
             ]

final_estimator = RFC(n_estimators=100
                      , min_impurity_decrease=0.0025
                      , random_state= 420, n_jobs=8)

clf = StackingClassifier(estimators=estimators
                         ,final_estimator=final_estimator
                         ,stack_method = "auto"
                         ,n_jobs=8)
 
clf = clf.fit(Xtrain,Ytrain)

In [34]:
# 接口transform来查看当前元学习器所使用的训练特征矩阵的结构
clf.transform(Xtrain).shape

(1437, 70)

In [35]:
clf = StackingClassifier(estimators=estimators
                         ,final_estimator=final_estimator
                         ,stack_method = "auto"
                         ,passthrough = True
                         ,n_jobs=8)
 
 
clf = clf.fit(Xtrain,Ytrain)

In [36]:
# 加入参数passthrough，特征矩阵的特征量会变得更大
clf.transform(Xtrain).shape

(1437, 134)

In [37]:
# 属性stack_method_，我们可以查看现在每个个体学习器都使用了什么接口做为预测输出
clf.stack_method_

'''
7个个体学习器都使用了predict_proba的概率接口进行输出，这与我们选择的算法都是可以输出概率的算法有很大的关系。 
'''

['predict_proba',
 'predict_proba',
 'predict_proba',
 'predict_proba',
 'predict_proba',
 'predict_proba',
 'predict_proba']

## Stacking融合的训练和测试流程

<center><img src="photos\stacking.png" alt="image-20220520122231347" style="zoom:83%;" />

### **stacking的训练**:

- 将数据分割为训练集、测试集，其中训练集上的样本为𝑀𝑡𝑟𝑎𝑖𝑛，测试集上的样本量为𝑀𝑡𝑒𝑠𝑡
 ---
- 将训练集输入level 0的个体学习器，分别在每个个体学习器上进行交叉验证。在每个个体学习器上，将所有交叉验证的验证结果纵向堆叠形成预测结果。假设预测结果为概率值，当融合模型执行回归或二分类任务时，该预测结果的结构为(𝑀𝑡𝑟𝑎𝑖𝑛,1)，当融合模型执行K分类任务时(K>2)，该预测结果的结构为(𝑀𝑡𝑟𝑎𝑖𝑛,𝐾)
 ---
- **隐藏步骤：使用全部训练数据对所有个体学习器进行训练，为测试做好准备。**
---
- 将所有个体学习器的预测结果横向拼接，形成新特征矩阵。假设共有N个个体学习器，当融合模型执行回归或二分类任务时，则新特征矩阵的结构为(𝑀𝑡𝑟𝑎𝑖𝑛,𝑁)。如果是输出多分类的概率，那最终得出的新特征矩阵的结构为(𝑀𝑡𝑟𝑎𝑖𝑛,𝑁∗𝐾)
---
- 将新特征矩阵放入元学习器进行训练。


### stacking的测试

- 将测试集输入level 0的个体学习器，分别在每个个体学习器上预测出相应结果。假设测试结果为概率值，当融合模型执行回归或二分类任务时，该测试结果的结构为(𝑀𝑡𝑒𝑠𝑡,1)，当融合模型执行K分类任务时(K>2)，该测试结果的结构为(𝑀𝑡𝑒𝑠𝑡,𝐾)
 ---
- 将所有个体学习器的预测结果横向拼接为新特征矩阵。假设共有N个个体学习器，则新特征矩阵的结构为(𝑀𝑡𝑒𝑠𝑡,𝑁)，如果是输出多分类的概率，那最终得出的新特征矩阵的结构为(𝑀𝑡𝑒𝑠𝑡, 𝑁∗𝐾)
---
- 将新特征矩阵放入元学习器进行预测。


因此在stacking中，不仅要对个体学习器完成全部交叉验证，还需要在交叉验证结束后，重新使用训练数据来训练所有的模型。无怪Stacking融合的复杂度较高、并且运行缓慢了

---

# Blending混合法
---
Blending融合是在Stacking融合的基础上改进过后的算法

**Blending直译为“混合”，但它的核心思路其实与Stacking完全一致：**

>使用两层算法串联，level 0上存在多个强学习器，level 1上有且只有一个元学习器，且level 0上的强学习器负责拟合数据与真实标签之间的关系、并输出预测结果、组成新的特征矩阵，然后让level 1上的元学习器在新的特征矩阵上学习并预测。


# 装袋算法Bagging
---
```Bagging```通过采样训练数据集的样本，训练得到多样的模型，进而得到多样的预测结果。

在结合模型的预测结果时，可以对单个模型预测结果进行投票或平均。
>Bagging的**关键**是对数据集的采样方法.常见的方式可以从行（样本）维度进行采样，这里进行的是有放回采样。

```Bagging```可通过```BaggingClassifier```和```BaggingRegressor```使用，默认情况下它们使用决策树作为基本模型，可以通过n_estimators参数指定要创建的树的数量。

|参数	|说明|
|----|----|
|base_estimator	|object, default=None 基本估计量适合数据集的随机子集。如果为None，则基本估计量为决策树。|
|n_estimators	|int，default = 10 集合中基本估计量的数量。|
|max_samples	|int or float, default=1.0 从X抽取以训练每个基本估计量的样本数量.如果为int，则抽取max_samples样本。;如果为float，则抽取样品。max_samples * X.shape[0]。|
|max_features	|int or float, default=1.0从X绘制以训练每个基本估计量的要素数量;如果为int，则绘制max_features特征。;如果为float，则绘制特征。max_features * X.shape[1]|
|bootstrap	|bool, default=True是否抽取样本进行替换。如果为False，则执行不替换的采样。|
|bootstrap_features	|bool, default=False是否用替换绘制特征。|
|oob_score|bool，defalut = False是否使用现成的样本来估计泛化误差。|
|warm_start|	bool，defalut = False设置为True时，请重用上一个调用的解决方案以适合并在集合中添加更多估计量，否则，仅适合一个全新的集合。请参阅Glossary。|
|n_jobs	|int, default=None (fit和 并行运行的作业数predict)|
|verbose	|int, default=0在拟合和预测时控制冗余程度。|

**原理过程**:

- 将原始数据集D随机划分成m个子集D1, D2, …, Dm
- 对于每个子集Di, 训练一个基学习器Hi
- 对于每个测试样本x，将其输入到所有的基学习器Hi中，并得到对应的预测结果{yi1, yi2, …, yim}
- 将所有的预测结果进行结合，以得出最终的预测结果y


<center><img src="photos\927391-20160717135005498-1140287801.png" alt="image-20220520122231347" style="zoom:83%;" />

## 优缺点
Bagging算法**优点**：

- Bagging算法可以显著降低模型的方差，提高模型的稳定性。通过使用多个相互独立的学习器，Bagging可以减少模型对训练数据的敏感性，从而更好地适应未知的测试数据。
- Bagging算法可以并行计算，加速模型训练的速度。
- Bagging算法不容易出现过拟合。通过使用随机抽样的方式来生成多个子集

Bagging算法的**缺点**包括：

- Bagging算法对于噪声数据比较敏感。由于Bagging算法对于每个子集都会训练一个基学习器，因此如果某个子集**包含大量的噪声数据**，那么对应的基学习器的性能可能会下降，从而影响整个模型的性能。
- Bagging算法在处理分类问题时，容易出现**过于一致性问题**。如果不同的基学习器都对同一个测试样本做出了**相同的错误预测**，那么Bagging算法将无法有效纠正这个错误。


In [3]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris

iris=load_iris()

X=iris.data[:,2:4]
y=iris.target

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2)

In [6]:
bgg_clf=BaggingClassifier(
    base_estimator=DecisionTreeClassifier(max_depth=2),
    n_estimators=10,
    bootstrap=True,
    random_state=42
)

bgg_clf.fit(X_train,y_train)

y_pred=bgg_clf.predict(X_test)

from sklearn.metrics import accuracy_score

print('准确率:',accuracy_score(y_test,y_pred))

准确率: 0.9666666666666667


C:\Users\36085\anaconda3\Lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


## OOB策略

**OOB（Out-Of-Bag）策略是Bagging算法的一种特殊形式，可以用于评估Bagging模型的性能和特征重要性**。

使用oob_score_属性来获取模型的OOB评估结果。使用OOB策略可以提高Bagging算法的性能，并在一定程度上避免了过拟合的问题。

同时，通过统计每个特征在OOB样本上的预测结果，我们还可以评估特征的重要性，从而进一步优化模型的性能。

需要注意的是，在使用OOB策略时，我们需要确保每个基学习器都使用了不同的子样本进行训练。如果使用相同的子样本进行训练，那么在OOB样本中可能会出现重复样本，导致评估结果不准确。


In [9]:
bgg_clf=BaggingClassifier(
    base_estimator=DecisionTreeClassifier(max_depth=2),
    n_estimators=10,
    bootstrap=True,
    random_state=42,
    oob_score=True
)

bgg_clf.fit(X_train,y_train)
bgg_clf.oob_score_

C:\Users\36085\anaconda3\Lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


0.9333333333333333

In [11]:
y_pred=bgg_clf.predict(X_test)
accuracy_score(y_test,y_pred)

0.9666666666666667

In [12]:
bgg_clf.oob_decision_function_     #表示每个类别所属的概率

array([[1.        , 0.        , 0.        ],
       [0.        , 0.04487179, 0.95512821],
       [0.        , 0.        , 1.        ],
       [0.        , 0.94425445, 0.05574555],
       [1.        , 0.        , 0.        ],
       [1.        , 0.        , 0.        ],
       [0.        , 0.01282051, 0.98717949],
       [0.        , 0.02106722, 0.97893278],
       [1.        , 0.        , 0.        ],
       [1.        , 0.        , 0.        ],
       [0.        , 0.01081081, 0.98918919],
       [0.        , 0.96376812, 0.03623188],
       [0.        , 0.01282051, 0.98717949],
       [1.        , 0.        , 0.        ],
       [0.        , 0.96956111, 0.03043889],
       [1.        , 0.        , 0.        ],
       [0.        , 0.93435635, 0.06564365],
       [0.        , 0.47471169, 0.52528831],
       [1.        , 0.        , 0.        ],
       [0.        , 0.0231982 , 0.9768018 ],
       [0.        , 0.01388889, 0.98611111],
       [0.        , 0.02358368, 0.97641632],
       [1.